### ISSUE TITLE
Why use itertools.chain() in discriminator and generator?

### ISSUE THREAD

**Issue by SadMemories** [junyanz/pytorch-CycleGAN-and-pix2pix#1381]
I was looking at ```cycle_gan_model.py``` and there, i noticed  that the following code uses itertools.chain() function to chain the parameters of both  discriminator and generator, my question is why use a single optimizer for both network. How does this work? Is it better than using two different optimizers for two different networks ?
```python
self.optimizer_G = torch.optim.Adam(itertools.chain(self.netG_A.parameters(), self.netG_B.parameters()), lr=opt.lr, betas=(opt.beta1, 0.999))
self.optimizer_D = torch.optim.Adam(itertools.chain(self.netD_A.parameters(), self.netD_B.parameters()), lr=opt.lr, betas=(opt.beta1, 0.999))
```

**Comment by junyanz** [comment]
You can use two different optimizers for two different networks. The effect should be the same. Using `chain` simplifies the code.

**Comment by SadMemories** [comment]
> 

Thank you for your answer, but i have one more question: if i use two different optimizers for netG_A and netG_B, when i calculate the cycle loss, do i have to call ``` backward() ``` on the losses of both generators before calling ``` step() ```? Or can I first calculate the loss of ```netG_A(netG_B)``` and update the gradient using ```step()```, and then calculate the loss of ```netG_B(netG_A)``` and update the gradient using ```step```? Which method should I use?

**Comment by junyanz** [comment]
I believe that you can call one `backward` (for the total loss) and two `step` (for netG_A and netG_B).

**Comment by sungjik** [comment]
Why is the effect the same when using one vs two different optimizers for each network?

**Comment by junyanz** [comment]
Even if you use chain, the parameters are independently updated. By "the same effect", I meant that it should not affect the final result of the training.
